cumulus/tasks/sync-granule/tests/sync_granule_test.js

```diff
@@ -98,6 +100,68 @@ test('download Granule from HTTP endpoint', (t) => {
     });
 });
 
+test('download Granule from S3 endpoint', (t) => {
+  const provider = {
+    id: 'MODAPS',
+    protocol: 's3'
+  };
+
+  const newPayload = Object.assign({}, payload);
+  newPayload.config.provider = provider;
+
+  // update path to start with s3://, upload the granule files to s3 for testing,
+  // update information of test files
+  for (let i = 0; i < newPayload.input.granules.length; i++) {
+    let granule = newPayload.input.granules[i];
+
+    for (let j = 0; j < granule.files.length; j++) {
+      newPayload.input.granules[i].files[j].path =
+        newPayload.input.granules[i].files[j].path.replace(/^\//, 's3://');
+
+      newPayload.input.granules[i].files[j].fileSize = 9;
+      newPayload.input.granules[i].files[j].checksumType = 'CKSUM';
+      newPayload.input.granules[i].files[j].checksumValue = 275331806;
+
+      let f = newPayload.input.granules[i].files[j];
+      const params = aws.parseS3Uri(`${f.path.replace(/\/+$/, '')}/${f.name}`);
+      aws.s3().createBucket({ Bucket: params.Bucket }).promise()
+        .then(() => aws.s3().putObject({
+          Bucket: params.Bucket,
+          Key: params.Key,
+          Body: 'test data'
+        }).promise()).then();
+    }
+  }
+
+  newPayload.input.pdr.path = newPayload.input.pdr.path.replace(/^\//, 's3://');
+
+  const protectedBucketName = testUtils.randomString();
+  const internalBucketName = testUtils.randomString();
+
+  newPayload.config.buckets.protected = protectedBucketName;
+  newPayload.config.buckets.internal = internalBucketName;
+
+  return aws.s3().createBucket({ Bucket: protectedBucketName }).promise()
+    .then(() => aws.s3().createBucket({ Bucket: internalBucketName }).promise())
+    .then(() => syncGranule(newPayload))
+    .then((output) => {
+      t.is(output.granules.length, 1);
+      t.is(output.granules[0].files.length, 1);
+      t.is(
+        output.granules[0].files[0].filename,
+        `s3://${protectedBucketName}/MOD09GQ.A2017224.h27v08.006.2017227165029.hdf`
+      );
+
+      return aws.recursivelyDeleteS3Bucket(internalBucketName);
+    })
+    .catch((e) => {
```

Comment:
Buckets are not deleted if there is an error.

Reply:
updated